### PR TITLE
Nooba SA deployment

### PR DESCRIPTION
**Pull request by vavuthu**
Signed-off-by: vavuthu [vavuthu@redhat.com](mailto:vavuthu@redhat.com)

### CODE REVIEW

**Comment by vavuthu** [comment]
verified
```
(venv2) [redhat@noobaasa noobaa-sa-ci]$ noobaa-sa-ci tests/deployment/test_deployment.py --conf packages.yaml
=========================================================================================== test session starts ============================================================================================
platform linux -- Python 3.8.16, pytest-7.4.2, pluggy-1.3.0
rootdir: /home/redhat/noobaa-sa-ci
configfile: pytest.ini
collected 1 item                                                                                                                                                                                           

tests/deployment/test_deployment.py::test_deployment 
---------------------------------------------------------------------------------------------- live log call -----------------------------------------------------------------------------------------------
2023-09-22 06:22:24 - MainThread - deployment.deployment - INFO - Installing Noobaa Standalone
2023-09-22 06:22:27 - MainThread - deployment.deployment - INFO - Downloaded RPM package to /tmp/noobaa-core-5.14.0-1-xs3select.el8.x86_64_vq9s_ao_.rpm
2023-09-22 06:22:27 - MainThread - utility.utils - INFO - Installing /tmp/noobaa-core-5.14.0-1-xs3select.el8.x86_64_vq9s_ao_.rpm
2023-09-22 06:22:27 - MainThread - utility.utils - INFO - Executing command: sudo yum install /tmp/noobaa-core-5.14.0-1-xs3select.el8.x86_64_vq9s_ao_.rpm -y
2023-09-22 06:22:46 - MainThread - utility.utils - INFO - Command stdout: Last metadata expiration check: 0:57:52 ago on Fri 22 Sep 2023 05:24:36 AM EDT.
Dependencies resolved.
================================================================================
 Package            Architecture  Version             Repository           Size
================================================================================
Installing:
 noobaa-core        x86_64        5.14.0-1.el8        @commandline         58 M

Transaction Summary
================================================================================
Install  1 Package

Total size: 58 M
Installed size: 375 M
Downloading Packages:
Running transaction check
Transaction check succeeded.
Running transaction test
Transaction test succeeded.
Running transaction
  Preparing        :                                                        1/1 
  Installing       : noobaa-core-5.14.0-1.el8.x86_64                        1/1 
  Running scriptlet: noobaa-core-5.14.0-1.el8.x86_64                        1/1 
  Verifying        : noobaa-core-5.14.0-1.el8.x86_64                        1/1 

Installed:
  noobaa-core-5.14.0-1.el8.x86_64                                               

Complete!

2023-09-22 06:22:46 - MainThread - utility.utils - INFO - Command stderr is empty
2023-09-22 06:22:46 - MainThread - utility.utils - INFO - Command return code: 0
2023-09-22 06:22:46 - MainThread - utility.utils - INFO - Successfully installed /tmp/noobaa-core-5.14.0-1-xs3select.el8.x86_64_vq9s_ao_.rpm
2023-09-22 06:22:46 - MainThread - utility.utils - INFO - Installing https://download.postgresql.org/pub/repos/yum/reporpms/EL-8-x86_64/pgdg-redhat-repo-latest.noarch.rpm
2023-09-22 06:22:46 - MainThread - utility.utils - INFO - Executing command: sudo yum install https://download.postgresql.org/pub/repos/yum/reporpms/EL-8-x86_64/pgdg-redhat-repo-latest.noarch.rpm -y
2023-09-22 06:22:49 - MainThread - utility.utils - INFO - Command stdout: Last metadata expiration check: 0:58:10 ago on Fri 22 Sep 2023 05:24:36 AM EDT.
pgdg-redhat-repo-latest.noarch.rpm               21 kB/s |  13 kB     00:00    
Package pgdg-redhat-repo-42.0-35PGDG.noarch is already installed.
Dependencies resolved.
Nothing to do.
Complete!

2023-09-22 06:22:49 - MainThread - utility.utils - INFO - Command stderr is empty
2023-09-22 06:22:49 - MainThread - utility.utils - INFO - Command return code: 0
2023-09-22 06:22:49 - MainThread - utility.utils - INFO - Successfully installed https://download.postgresql.org/pub/repos/yum/reporpms/EL-8-x86_64/pgdg-redhat-repo-latest.noarch.rpm
2023-09-22 06:22:49 - MainThread - utility.utils - INFO - Enable the module stream for Postgres version to 12
2023-09-22 06:22:49 - MainThread - utility.utils - INFO - Executing command: sudo dnf module enable postgresql:12
2023-09-22 06:22:51 - MainThread - utility.utils - INFO - Command stdout: Last metadata expiration check: 0:58:13 ago on Fri 22 Sep 2023 05:24:36 AM EDT.
Dependencies resolved.
Nothing to do.
Complete!

2023-09-22 06:22:51 - MainThread - utility.utils - INFO - Command stderr is empty
2023-09-22 06:22:51 - MainThread - utility.utils - INFO - Command return code: 0
2023-09-22 06:22:51 - MainThread - utility.utils - INFO - Installing libpq postgresql postgresql-server
2023-09-22 06:22:51 - MainThread - utility.utils - INFO - Executing command: sudo yum install libpq postgresql postgresql-server -y
2023-09-22 06:24:06 - MainThread - utility.utils - INFO - Command stdout: Last metadata expiration check: 0:58:16 ago on Fri 22 Sep 2023 05:24:36 AM EDT.
Package libpq-13.5-1.el8.x86_64 is already installed.
Dependencies resolved.
================================================================================
 Package           Arch   Version                               Repo       Size
================================================================================
Installing:
 postgresql        x86_64 12.12-1.module_el8.6.0+1219+2bd8590a  appstream 1.5 M
 postgresql-server x86_64 12.12-1.module_el8.6.0+1219+2bd8590a  appstream 5.6 M

Transaction Summary
================================================================================
Install  2 Packages

Total download size: 7.1 M
Installed size: 28 M
Downloading Packages:
(1/2): postgresql-12.12-1.module_el8.6.0+1219+2 102 kB/s | 1.5 MB     00:14    
(2/2): postgresql-server-12.12-1.module_el8.6.0  80 kB/s | 5.6 MB     01:10    
--------------------------------------------------------------------------------
Total                                           102 kB/s | 7.1 MB     01:11     
Running transaction check
Transaction check succeeded.
Running transaction test
Transaction test succeeded.
Running transaction
  Preparing        :                                                        1/1 
  Installing       : postgresql-12.12-1.module_el8.6.0+1219+2bd8590a.x86_   1/2 
  Running scriptlet: postgresql-server-12.12-1.module_el8.6.0+1219+2bd859   2/2 
  Installing       : postgresql-server-12.12-1.module_el8.6.0+1219+2bd859   2/2 
  Running scriptlet: postgresql-server-12.12-1.module_el8.6.0+1219+2bd859   2/2 
  Verifying        : postgresql-12.12-1.module_el8.6.0+1219+2bd8590a.x86_   1/2 
  Verifying        : postgresql-server-12.12-1.module_el8.6.0+1219+2bd859   2/2 

Installed:
  postgresql-12.12-1.module_el8.6.0+1219+2bd8590a.x86_64                        
  postgresql-server-12.12-1.module_el8.6.0+1219+2bd8590a.x86_64                 

Complete!

2023-09-22 06:24:06 - MainThread - utility.utils - INFO - Command stderr is empty
2023-09-22 06:24:06 - MainThread - utility.utils - INFO - Command return code: 0
2023-09-22 06:24:06 - MainThread - utility.utils - INFO - Successfully installed None
2023-09-22 06:24:06 - MainThread - utility.utils - INFO - Executing command: sudo chmod 777 /usr/local/noobaa-core
2023-09-22 06:24:06 - MainThread - utility.utils - INFO - Command stdout is empty
2023-09-22 06:24:06 - MainThread - utility.utils - INFO - Command stderr is empty
2023-09-22 06:24:06 - MainThread - utility.utils - INFO - Command return code: 0
2023-09-22 06:24:06 - MainThread - utility.utils - INFO - Permissions for '/usr/local/noobaa-core' have been set to 777
2023-09-22 06:24:06 - MainThread - utility.utils - INFO - Directory '/usr/local/noobaa-core/storage' created successfully.
2023-09-22 06:24:06 - MainThread - utility.utils - INFO - Executing command: sudo chmod 777 /var/run/postgresql
2023-09-22 06:24:06 - MainThread - utility.utils - INFO - Command stdout is empty
2023-09-22 06:24:06 - MainThread - utility.utils - INFO - Command stderr is empty
2023-09-22 06:24:06 - MainThread - utility.utils - INFO - Command return code: 0
2023-09-22 06:24:06 - MainThread - utility.utils - INFO - Permissions for '/var/run/postgresql' have been set to 777
2023-09-22 06:24:06 - MainThread - deployment.deployment - INFO - Initializing DB
2023-09-22 06:24:06 - MainThread - deployment.npm - INFO - executing 'npm run db:init'

> noobaa-core@5.14.0 db:init
> initdb -D ./storage/postgres -E UTF-8 --locale=C

The files belonging to this database system will be owned by user "redhat".
This user must also own the server process.

The database cluster will be initialized with locale "C".
The default text search configuration will be set to "english".

Data page checksums are disabled.

creating directory ./storage/postgres ... ok
creating subdirectories ... ok
selecting dynamic shared memory implementation ... posix
selecting default max_connections ... 100
selecting default shared_buffers ... 128MB
selecting default time zone ... America/New_York
creating configuration files ... ok
running bootstrap script ... ok
performing post-bootstrap initialization ... ok
syncing data to disk ... ok

initdb: warning: enabling "trust" authentication for local connections
You can change this by editing pg_hba.conf or using the option -A, or
--auth-local and --auth-host, the next time you run initdb.

Success. You can now start the database server using:

    pg_ctl -D ./storage/postgres -l logfile start

2023-09-22 06:24:07 - MainThread - deployment.deployment - INFO - starting the DB
2023-09-22 06:24:07 - MainThread - deployment.npm - INFO - executing 'npm run db'
2023-09-22 06:24:17 - MainThread - deployment.deployment - INFO - Creating DB, users and permissions
2023-09-22 06:24:17 - MainThread - deployment.npm - INFO - executing 'npm run db:create'

> noobaa-core@5.14.0 db:create
> npm run db:createuser && npm run db:createdb


> noobaa-core@5.14.0 db:createuser
> createuser postgres && psql postgres -c 'ALTER USER postgres CREATEDB'

ALTER ROLE

> noobaa-core@5.14.0 db:createdb
> createdb -U postgres nbcore

2023-09-22 06:24:28 - MainThread - deployment.deployment - INFO - creating .env file
2023-09-22 06:24:28 - MainThread - utility.utils - INFO - Executing command: sudo touch /usr/local/noobaa-core/.env
2023-09-22 06:24:28 - MainThread - utility.utils - INFO - Command stdout is empty
2023-09-22 06:24:28 - MainThread - utility.utils - INFO - Command stderr is empty
2023-09-22 06:24:28 - MainThread - utility.utils - INFO - Command return code: 0
2023-09-22 06:24:28 - MainThread - utility.utils - INFO - Executing command: sudo chmod 666 /usr/local/noobaa-core/.env
2023-09-22 06:24:28 - MainThread - utility.utils - INFO - Command stdout is empty
2023-09-22 06:24:28 - MainThread - utility.utils - INFO - Command stderr is empty
2023-09-22 06:24:28 - MainThread - utility.utils - INFO - Command return code: 0
2023-09-22 06:24:28 - MainThread - deployment.deployment - INFO - creating config-local.js file
2023-09-22 06:24:28 - MainThread - utility.utils - INFO - Executing command: sudo cp /home/redhat/noobaa-sa-ci/templates/config-local.js /usr/local/noobaa-core/config-local.js
2023-09-22 06:24:28 - MainThread - utility.utils - INFO - Command stdout is empty
2023-09-22 06:24:28 - MainThread - utility.utils - INFO - Command stderr is empty
2023-09-22 06:24:28 - MainThread - utility.utils - INFO - Command return code: 0
2023-09-22 06:24:28 - MainThread - deployment.deployment - INFO - starting the web service
2023-09-22 06:24:28 - MainThread - deployment.npm - INFO - executing 'npm run web'
2023-09-22 06:24:38 - MainThread - deployment.deployment - INFO - starting the bg service
2023-09-22 06:24:38 - MainThread - deployment.npm - INFO - executing 'npm run bg'
2023-09-22 06:24:48 - MainThread - deployment.deployment - INFO - starting the hosted agent service
2023-09-22 06:24:48 - MainThread - deployment.npm - INFO - executing 'npm run hosted_agents'
2023-09-22 06:24:58 - MainThread - deployment.deployment - INFO - starting the s3 endpoint service
2023-09-22 06:24:58 - MainThread - deployment.npm - INFO - executing 'npm run s3'
2023-09-22 06:25:08 - MainThread - utility.utils - INFO - Directory '/usr/local/noobaa-core/storage/backingstores/drive0' created successfully.
2023-09-22 06:25:08 - MainThread - deployment.deployment - INFO - running backing store '/usr/local/noobaa-core/storage/backingstores/drive0' at port 9991
2023-09-22 06:25:08 - MainThread - deployment.npm - INFO - executing 'npm run backingstore'
2023-09-22 06:25:18 - MainThread - utility.utils - INFO - Directory '/usr/local/noobaa-core/storage/backingstores/drive1' created successfully.
2023-09-22 06:25:18 - MainThread - deployment.deployment - INFO - running backing store '/usr/local/noobaa-core/storage/backingstores/drive1' at port 9992
2023-09-22 06:25:18 - MainThread - deployment.npm - INFO - executing 'npm run backingstore'
2023-09-22 06:25:28 - MainThread - utility.utils - INFO - Directory '/usr/local/noobaa-core/storage/backingstores/drive2' created successfully.
2023-09-22 06:25:28 - MainThread - deployment.deployment - INFO - running backing store '/usr/local/noobaa-core/storage/backingstores/drive2' at port 9993
2023-09-22 06:25:28 - MainThread - deployment.npm - INFO - executing 'npm run backingstore'
2023-09-22 06:25:38 - MainThread - utility.utils - INFO - Directory '/usr/local/noobaa-core/storage/backingstores/drive3' created successfully.
2023-09-22 06:25:38 - MainThread - deployment.deployment - INFO - running backing store '/usr/local/noobaa-core/storage/backingstores/drive3' at port 9994
2023-09-22 06:25:38 - MainThread - deployment.npm - INFO - executing 'npm run backingstore'
2023-09-22 06:25:48 - MainThread - deployment.deployment - INFO - checking storage status
2023-09-22 06:25:48 - MainThread - deployment.npm - INFO - executing 'npm run api'

> noobaa-core@5.14.0 api
> node --unhandled-rejections=warn src/cmd/api node sync_monitor_to_store

load_config_local: LOADED { DEFAULT_POOL_TYPE: 'HOSTS', AGENT_RPC_PORT: '9999', AGENT_RPC_PROTOCOL: 'tcp', ENDPOINT_HTTP_SERVER_REQUEST_TIMEOUT: 900000, ENDPOINT_HTTP_SERVER_KEEPALIVE_TIMEOUT: 60000, BUCKET_AUTOCONF_TIER2_ENABLED: true, BLOCK_STORE_FS_TMFS_ENABLED: true, BLOCK_STORE_FS_MAPPING_INFO_ENABLED: true, DEDUP_ENABLED: false, IO_CALC_MD5_ENABLED: false, IO_CALC_SHA256_ENABLED: false, MAX_OBJECT_PART_SIZE: 1073741824, IO_CHUNK_READ_CACHE_SIZE: 1073741824, IO_READ_BLOCK_TIMEOUT: 600000, IO_WRITE_BLOCK_TIMEOUT: 600000, IO_DELETE_BLOCK_TIMEOUT: 600000, IO_REPLICATE_BLOCK_TIMEOUT: 600000, NODE_IO_DETENTION_DISABLE: true, NODE_IO_DETENTION_THRESHOLD: 0, CHUNK_SPLIT_AVG_CHUNK: 268435456, CHUNK_SPLIT_DELTA_CHUNK: 0, CHUNK_CODER_DIGEST_TYPE: 'none', CHUNK_CODER_FRAG_DIGEST_TYPE: 'none', CHUNK_CODER_COMPRESS_TYPE: 'none', CHUNK_CODER_CIPHER_TYPE: 'none', CHUNK_CODER_REPLICAS: 1, CHUNK_CODER_EC_DATA_FRAGS: 2, CHUNK_CODER_EC_PARITY_FRAGS: 2, CHUNK_CODER_EC_PARITY_TYPE: 'cm256', CHUNK_CODER_EC_TOLERANCE_THRESHOLD: 2, CHUNK_CODER_EC_IS_DEFAULT: true, SCRUBBER_ENABLED: false, REBUILD_NODE_ENABLED: false, AWS_METERING_ENABLED: false, AGENT_BLOCKS_VERIFIER_ENABLED: false, TIERING_TTL_WORKER_ENABLED: true }
undefined
2023-09-22 06:25:54 - MainThread - deployment.deployment - INFO - Storage Check successfully passed
2023-09-22 06:25:54 - MainThread - deployment.deployment - INFO - checking aggregate node status
2023-09-22 06:25:54 - MainThread - deployment.npm - INFO - executing 'npm run api'

> noobaa-core@5.14.0 api
> node --unhandled-rejections=warn src/cmd/api node aggregate_nodes {}

load_config_local: LOADED { DEFAULT_POOL_TYPE: 'HOSTS', AGENT_RPC_PORT: '9999', AGENT_RPC_PROTOCOL: 'tcp', ENDPOINT_HTTP_SERVER_REQUEST_TIMEOUT: 900000, ENDPOINT_HTTP_SERVER_KEEPALIVE_TIMEOUT: 60000, BUCKET_AUTOCONF_TIER2_ENABLED: true, BLOCK_STORE_FS_TMFS_ENABLED: true, BLOCK_STORE_FS_MAPPING_INFO_ENABLED: true, DEDUP_ENABLED: false, IO_CALC_MD5_ENABLED: false, IO_CALC_SHA256_ENABLED: false, MAX_OBJECT_PART_SIZE: 1073741824, IO_CHUNK_READ_CACHE_SIZE: 1073741824, IO_READ_BLOCK_TIMEOUT: 600000, IO_WRITE_BLOCK_TIMEOUT: 600000, IO_DELETE_BLOCK_TIMEOUT: 600000, IO_REPLICATE_BLOCK_TIMEOUT: 600000, NODE_IO_DETENTION_DISABLE: true, NODE_IO_DETENTION_THRESHOLD: 0, CHUNK_SPLIT_AVG_CHUNK: 268435456, CHUNK_SPLIT_DELTA_CHUNK: 0, CHUNK_CODER_DIGEST_TYPE: 'none', CHUNK_CODER_FRAG_DIGEST_TYPE: 'none', CHUNK_CODER_COMPRESS_TYPE: 'none', CHUNK_CODER_CIPHER_TYPE: 'none', CHUNK_CODER_REPLICAS: 1, CHUNK_CODER_EC_DATA_FRAGS: 2, CHUNK_CODER_EC_PARITY_FRAGS: 2, CHUNK_CODER_EC_PARITY_TYPE: 'cm256', CHUNK_CODER_EC_TOLERANCE_THRESHOLD: 2, CHUNK_CODER_EC_IS_DEFAULT: true, SCRUBBER_ENABLED: false, REBUILD_NODE_ENABLED: false, AWS_METERING_ENABLED: false, AGENT_BLOCKS_VERIFIER_ENABLED: false, TIERING_TTL_WORKER_ENABLED: true }
{
  nodes: { count: 4, online: 4, by_mode: { OPTIMAL: 3, INITIALIZING: 1 } },
  storage_nodes: { count: 4, online: 4, by_mode: { OPTIMAL: 3, INITIALIZING: 1 } },
  storage: {
    total: 300500910080,
    free: 199964172288,
    used: 0,
    reserved: 419430400,
    unavailable_free: 66654724096,
    unavailable_used: 0,
    used_other: 33462583296
  },
  data_activities: [ [length]: 0 ]
}
2023-09-22 06:25:55 - MainThread - deployment.deployment - INFO - Aggregate node status check successfully passed
PASSED

====================================================================================== 1 passed in 210.96s (0:03:30) =======================================================================================
(venv2) [redhat@noobaasa noobaa-sa-ci]$ 

```

File: deployment/deployment.py [hunk 1]
```diff
@@ -0,0 +1,307 @@
...
+        # check storage status
+        self.check_storage_status()
+
+        # check node status
+        self.check_node_status()
+
+        # switch to original directory
+        os.chdir(previous_dir)
+
+    def download_rpms(self, rpm_url):
```
Comment (petr-balogh) [inline deployment/deployment.py:142]: It downloading one RPM, so I think we should just name it to `download_rpm`

File: framework/default_config.yaml [hunk 2]
```diff
@@ -0,0 +1,18 @@
+---
+RUN:
+  log_dir: "/tmp"
+  run_id: null  # this will be redefined in the execution
```
Comment (petr-balogh) [inline framework/default_config.yaml:4]: We don't changing the run_id anywhere or I missed it?

File: framework/default_config.yaml [hunk 3]
```diff
@@ -0,0 +1,18 @@
+---
+RUN:
+  log_dir: "/tmp"
```
Comment (petr-balogh) [inline framework/default_config.yaml:3]: Do we use somewhere this log_dir?

File: framework/default_config.yaml [hunk 4]
```diff
@@ -0,0 +1,18 @@
...
+  run_id: null  # this will be redefined in the execution
+  cli_params: {}  # this will be filled with CLI parameters data
```
Comment (petr-balogh) [inline framework/default_config.yaml:5]: Also don't see usage of it.

**Comment by dahorak** [comment]
Would it make sense (and would it be possible) to separate some of the really common functionality from _ocs-ci_ (e.g. _Templating_, _exec_cmd_, _create_directory_, _download_file_,...) to separated repository and reuse it both in _ocs-ci_ and in the Noobaa SA projects?

**Comment by vavuthu** [comment]
closing this as deployment code went to noobaa-sa-infra

**Pull request closed without merging**